lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1597,23 +1584,23 @@ static void checkGenericParamAccessibility(TypeChecker &TC,
     AEK_Parameter = 0,
     AEK_Requirement
   } accessibilityErrorKind;
-  const DeclContext *minAccessScope = nullptr;
+  auto minAccessScope = AccessScope::getPublic();
   const TypeRepr *complainRepr = nullptr;
 
   // Swift 3.0.0 mistakenly didn't diagnose any issues when the context access
   // scope represented a private or fileprivate level.
-  bool shouldDowngradeToWarning = accessScope && !isa<ModuleDecl>(accessScope);
+  bool shouldDowngradeToWarning =
+    !accessScope.isPublic() && !isa<ModuleDecl>(accessScope.getDeclContext());
 
   for (auto param : *params) {
     if (param->getInherited().empty())
       continue;
     assert(param->getInherited().size() == 1);
     checkTypeAccessibilityImpl(TC, param->getInherited().front(), accessScope,
                                owner->getDeclContext(),
-                               [&](const DeclContext *typeAccessScope,
+                               [&](AccessScope typeAccessScope,
                                    const TypeRepr *thisComplainRepr) {
-      if (!minAccessScope ||
-          typeAccessScope->isChildContextOf(minAccessScope) ||
+      if (typeAccessScope.isChildOf(minAccessScope) ||
           (!complainRepr && typeAccessScope == minAccessScope)) {
```

Comment:
I think this one's supposed to be checking for the same DC rather than identical scopes. ("If we can warn about either one, we should pick the one with a TypeRepr.")